File: R/custom-controls.R
#' Add a text control to the map
#' @inheritParams add_control
#' @param text The text (HTML) that is displayed.
#' @param css_text A \code{cssText} string that will modify the style of the control element.
#' @export
add_text_control <- function(map, text, pos = NULL, css_text = NULL) {
  options <- list(
    text = text,
    cssText = css_text
  )
  invoke_method(map, "addCustomControl", controlName = "TextControl", pos = pos, options = options)
}

#' Add a mouse position control to the map
#' @inheritParams add_text_control
#' @param mustache_template A \href{https://github.com/janl/mustache.js}{mustache} template
#'   that contains the mustache tags \code{lng} and \code{lat}.
#' @example examples/api-reference/mouse-position-control.R
#' @export
add_mouse_position_control <- function(map, mustache_template = NULL, pos = NULL, css_text = NULL) {
  options <- list(
    mustacheTemplate = mustache_template,
    cssText = css_text
  )
  invoke_method(map, "addCustomControl", controlName = "MousePositionControl", pos = pos, options = options)
}

#' Add a filter control to the map
#' @inheritParams add_text_control
#' @param layer_id The ID of the layer to which the filter is attached.
#' @param filter The initial filter expression.
#' @param rows The number of rows of the textarea input.
#' @param cols The number of columns of the textarea input.
#' @export
add_filter_control <- function(map, layer_id, filter = NULL, pos = NULL, rows = 1, cols = 20) {
  options <- list(
    layer = layer_id,
    filter = filter,
    textareaAttributes = list(rows = rows, cols = cols)
  )
  invoke_method(map, "addCustomControl", controlName = "FilterControl", pos = pos, options = options)
}

File: javascript/src/custom-controls.js
import { render } from "mustache";

class TextControl {
  constructor(options) {
    this._options = options || { };
  }

  onAdd(map) {
    this._map = map;
    this._container = document.createElement("div");
    this._container.classList.add("mapboxgl-ctrl", "mapboxer-text-ctrl");
    this._container.style.cssText = this._options.cssText || "";
    this._container.innerHTML = this._options.text || "Hello mapboxer!";
    return this._container;
  }

  onRemove() {
    this._container.parentNode.removeChild(this._container);
    this._map = undefined;
  }
}

class MousePositionControl {
  constructor(options) {
    this._options = options || { };
  }

  onAdd(map) {
    this._map = map;
    this._container = document.createElement("div");
    this._container.classList.add("mapboxgl-ctrl", "mapboxer-mouse-position-ctrl");
    this._container.style.cssText = this._options.cssText || "";
    const mustacheTemplate = this._options.mustacheTemplate || "{{lng}}, {{lat}}";
    this._map.on("mousemove", e => {
      this._container.innerHTML = render(mustacheTemplate, e.lngLat);
    });
    this._map.on("mouseout", e => {
      this._container.innerHTML = "";
      this._container.style.display = "none";
    });
    this._map.on("mouseover", e => {
      this._container.style.display = "block";
    });
    return this._container;
  }

  onRemove() {
    this._container.parentNode.removeChild(this._container);
    this._map = undefined;
  }
}

class FilterControl {
  constructor(options) {
    this._options = options || { };
  }

  onAdd(map) {
    this._map = map;
    this._container = document.createElement("div");
    this._container.classList.add("mapboxgl-ctrl", "mapboxer-filter-ctrl");
    this._container.style.cssText = this._options.cssText || "";
    // this._container.innerText = "Filter: ";

    // const input = document.createElement("input");
    // Object.assign(input, { type: "text", spellcheck: false, id: "filter" });
    const input = document.createElement("textarea");
    Object.assign(input, { spellcheck: false, id: "filter", rows: 5 });
    if (this._options.filter) {
      input.value = JSON.stringify(this._options.filter);
      map.on("load", () => map.setFilter(this._options.layer, this._options.filter));
    }

    input.addEventListener("keyup", e => {
      const expressionString = input.value;
      try {
        const expression = JSON.parse(expressionString);
        map.setFilter(this._options.layer, expression);
      } catch(err) {
        // console.log(err);
      }
    });
    this._container.append(input);
    return this._container;
  }

  onRemove() {
    this._container.parentNode.removeChild(this._container);
    this._map = undefined;
  }
}

export default {
  TextControl,
  MousePositionControl,
  FilterControl
};
